tmt/base.py

```diff
@@ -1346,3 +1371,76 @@ def export(self):
         if self.note:
             data['note'] = self.note
         return data
+
+
+class Link(object):
+    """ Core attribute link parsing """
+
+    _relations = [
+        'verifies', 'verified-by',
+        'implements', 'implemented-by',
+        'documents', 'documented-by',
+        'blocks', 'blocked-by',
+        'duplicates', 'duplicated-by',
+        'parent', 'child',
+        'relates',
+        ]
+
+    def __init__(self, data=None):
+        """ Convert link data into canonical form """
+        # Nothing to do if no data provided
+        self.links = []
+        if data is None:
+            return
+        if not isinstance(data, list):
+            data = [data]
+
+        # Ensure that each link is in the canonical form
+        for link in data:
+
+            # It should be a string
+            if isinstance(link, str):
+                self.links.append(dict(relates=link))
+                continue
+
+            # Or a dictionary
+            if not isinstance(link, dict):
+                raise tmt.utils.SpecificationError(
+                    f"Invalid link target (should be 'str' or 'dict', "
+                    f"got '{type(link).__name__}'.")
+
+            # Verify the relation
+            relations = []
+            for key in link:
+                if key in ['url', 'ref', 'path', 'name', 'note']:
+                    continue
+                if key in self._relations:
+                    relations.append(key)
+                    continue
+                raise tmt.utils.SpecificationError(
+                    f"Invalid link relation '{key}' (should be "
+                    f"{fmf.utils.listed(self._relations, join='or')}).")
+            # More relations (error)
+            if len(relations) > 1:
+                raise tmt.utils.SpecificationError(
+                    f"Multiple relations specified for the link "
+                    f"({fmf.utils.listed(relations)}).")
+            # No relation (fmf id)
+            if len(relations) == 0:
+                self.links.append(dict(relates=link))
+            # Exactly one relation (valid link)
+            else:
+                self.links.append(link)
+
+    def get(self, relation=None):
+        """ Get links with given relation, all by default """
+        return [
+            link for link in self.links
+            if relation in link or relation is None]
+
+    def show(self):
+        """ Format a list of links with their relations """
+        for link in self.links:
+            relation = [key for key in link.keys() if key != 'note'][0]
+            echo(tmt.utils.format(
+                relation.rstrip('-by'), f"{link[relation]}", key_color='cyan'))
```

Comment:
I'm not sure about `rstrip('-by')`:
- I was surprised to see different relation than I specified in the metadata
- You can't copy&paste name of the relation from show() output - you need to consult documentation (or fix after you see the fatal error)
- You have to read until the very end of the string to see what it really is. e.g. `verifies` vs `verified`

But I see it is more pleasant for the eye.

Reply:
Yeah, I spent some time pondering on this. Tried something like this:

```
/stories/implemented
     summary Story example
       story As a user I want to store relevant links.
        link implemented /code/example
```

But did not like it much. Then increased the left padding to make the longest relation fit:

```
/stories/implemented
        summary Story example
          story As a user I want to store relevant links.
 implemented-by /code/example
```

And then ended up with stripping the `-by` part because the large default padding seemed ugly. Not sure.

Reply:
I guess it should be OK for now. Tested on users ;)